### PR TITLE
Domains: Route to new email upsell page after adding domain to cart

### PR DESCRIPTION
**Pull request by runofthemillgeek**
#### Changes proposed in this Pull Request

This PR enables the new professional email upsell page built in #55183 by adding a feature flag to enable this feature and routing to this new page when a domain is added to cart, instead of the current GSuite upsell page.

#### Testing instructions

1. Checkout this branch: `add/domains-titan-email-routing-feature-flag` and start the server
2. Visit http://calypso.localhost:3000/domains/add?flags=titan/upsell-email-domain-purchase
3. Choose a site
4. Click _Select_ button next to a domain to add it to cart
5. You'll be taken to the new route.
6. Fill details for professional email/GSuite and click the corresponding _Add_ button
7. You'll be taken to the checkout screen

#### Recordings

<details>
<summary>
Click/tap here to see the recording

https://user-images.githubusercontent.com/2759499/129955322-983145f8-5a8b-4e32-8939-74054b9d8a94.mp4

</summary>
</details>

CC @daledupreez

### CODE REVIEW

File: client/my-sites/domains/domain-search/index.jsx [hunk 1]
```diff
@@ -155,6 +156,11 @@ class DomainSearch extends Component {
 				fillInSingleCartItemAttributes( registration, this.props.productsList ),
 			] )
 			.then( () => {
+				if ( config.isEnabled( 'titan/upsell-email-domain-purchase' ) ) {
+					page( domainAddEmailUpsell( this.props.selectedSiteSlug, domain ) );
+					return;
+				}
+
 				if ( this.props.userCanPurchaseGSuite && canDomainAddGSuite( domain ) ) {
```
Comment (runofthemillgeek) [inline client/my-sites/domains/domain-search/index.jsx]: @daledupreez you mentioned removing GSuite-specific logic on the Asana task but what needs to be done about this particular condition? Should this condition be moved into the `<EmailProvidersComparison>` component and the GSuite card conditionally rendered based on its truthiness? 🤔

File: client/my-sites/domains/paths.js [hunk 2]
```diff
@@ -61,7 +61,7 @@ export function domainAddNew( siteName, searchTerm ) {
 }
 
 export function domainAddEmailUpsell( siteName, domainName ) {
-	return `/domains/add/${ siteName }/email/${ domainName }`;
+	return `/domains/add/${ domainName }/email/${ siteName }`;
```
Comment (runofthemillgeek) [inline client/my-sites/domains/paths.js:61]: Also wanted to point out that I got the order of variables in template literal wrong in the original PR 🙈

Fixed it here.

File: client/my-sites/domains/domain-search/index.jsx (hunk 1 above)
Comment (daledupreez) [reply client/my-sites/domains/domain-search/index.jsx]: This logic is already embedded in `EmailProvidersComparison`:
https://github.com/Automattic/wp-calypso/blob/94c00e90e144208f73488287905df71e95e63dad/client/my-sites/email/email-providers-comparison/index.jsx#L694

However, I do notice that logic needs to be modified slightly due to us now not having a `domain` object coming from this upsell path. I think that code needs to be updated to the following:
```javascript
const isGSuiteSupported = canUserPurchaseGSuite( state ) && ( hasCartDomain || ( domain && hasGSuiteSupportedDomain( [ domain ] ) ) ); 
```

File: client/my-sites/domains/paths.js (hunk 2 above)
Comment (daledupreez) [reply client/my-sites/domains/paths.js:61]: Thanks for catching this! 😳

File: config/client.json [hunk 3]
```diff
@@ -31,6 +31,7 @@
 	"siftscience_key",
 	"signup_url",
 	"support-user",
+	"titan/upsell-email-domain-purchase",
```
Comment (daledupreez) [inline config/client.json]: You don't need to include the feature flag in this list.
```suggestion
```

File: config/development.json [hunk 4]
```diff
@@ -152,6 +152,7 @@
 		"site-indicator": true,
 		"memberships": true,
 		"support-user": true,
+		"titan/upsell-email-domain-purchase": false,
```
Comment (daledupreez) [inline config/development.json]: I have two thoughts that contradict each other here!

1. I don't know that we need a feature flag for this -- I think it's a simple enough change that we can just remove the routing quickly via a new PR. We can control the launch by waiting to get this PR deployed. This shouldn't be so disruptive that we need really extensive testing etc.
2. If we _really_ want a feature flag, I think we can just prefix it with `email`, as the changes here do still promote Google Workspace.

**Review by daledupreez — COMMENTED** [review]
I haven't tested yet as the PR is still a draft, but I did share some feedback. LMK if you'd like a full review with testing.

File: client/my-sites/domains/domain-search/index.jsx (hunk 1 above)
Comment (runofthemillgeek) [reply client/my-sites/domains/domain-search/index.jsx]: Thanks for the feedback! I'll test this out.

File: config/client.json (hunk 3 above)
Comment (runofthemillgeek) [reply config/client.json]: Is this section outdated/incorrect? https://github.com/Automattic/wp-calypso/blob/94c00e90e144208f73488287905df71e95e63dad/config/README.md#feature-flags

File: config/development.json (hunk 4 above)
Comment (runofthemillgeek) [reply config/development.json]: 1. It's a small enough change, I agree. Will have to clean up the flags later as well if we go this route. I'm fine either way. 😝 Let me know if I should remove the flag.
2. 💯

File: client/my-sites/domains/domain-search/index.jsx (hunk 1 above)
Comment (runofthemillgeek) [reply client/my-sites/domains/domain-search/index.jsx]: I've made the change and tested it by returning `true`/`false`.

File: client/my-sites/domains/domain-search/index.jsx [hunk 5]
```diff
@@ -145,6 +146,11 @@ class DomainSearch extends Component {
 				fillInSingleCartItemAttributes( registration, this.props.productsList ),
 			] )
 			.then( () => {
+				if ( config.isEnabled( 'titan/upsell-email-domain-purchase' ) ) {
+					page( domainAddEmailUpsell( this.props.selectedSiteSlug, domain ) );
+					return;
+				}
```
Comment (daledupreez) [inline client/my-sites/domains/domain-search/index.jsx]: I think we can drop the feature flag and just add this in directly when this gets merged.
(This is closely linked to [this conversation](https://github.com/Automattic/wp-calypso/pull/55592/files#r692071323).)

**Review by daledupreez — APPROVED** [review]
This LGTM!

I think we can simplify this PR by removing the feature flag, and just deferring the switch-over to when we're ready.

File: client/my-sites/domains/domain-search/index.jsx (hunk 5 above)
Comment (runofthemillgeek) [reply client/my-sites/domains/domain-search/index.jsx]: Have made the changes. No flag it is! 💯

**Comment by daledupreez** [comment]
I've retested and this LGTM.

**Comment by worldomonation** [comment]
@daledupreez @sangeeth96 This PR broke the following test in Selenium E2E:

- `Select .com search result and decline Google Apps offer`

For both desktop and mobile.

**Comment by worldomonation** [comment]
The step after adding to cart a paid domain, where the G Suite upsell would have previously occurred, is failing as the `gsuite-upsell-card__skip-button` no longer exists, being replaced with `email-providers-comparison__titan-mailbox-action-skip`.

I can push up a PR to fix this selector. For future reference, E2E tests for open-source contributors are not run by default; they must be manually triggered before merge to ensure changes do not affect Calypso negatively.

The following E2E tests should be triggered for Calypso prior to merge:
- Selenium E2E (mobile and desktop)
- Playwright E2E (mobile and desktop)

**Comment by daledupreez** [comment]
Thanks for identifying and correcting the issue, @worldomonation!

I re-read the instructions I followed and see that there is a gap in our instructions for verifying OSS contributions, which is how this problem slipped through.

**Pull request merged**